getgather/parse.py

```diff
@@ -117,64 +118,201 @@ async def _get_value(brand_id: BrandIdEnum, column: Column, element: Locator) ->
+                        value = ""
+                else:
+                    value = elem.get_text(strip=True)
+                row_data[column.name] = value if value else ""
+
+        data.append(row_data)
+
+    return data
+
+
 async def parse_html(
     brand_id: BrandIdEnum,
     schema: Schema,
     *,
     bundle_dir: Path | None = None,
     html_content: str | None = None,
+    page: Page | None = None,
 ) -> BundleOutput:
     """
-    Use headless Playwright to parse HTML content into a tabular format stored as
-    JSON file.
-    Assuming schema.row_selector is a valid CSS selector to identify the individual rows,
-    and schema.columns is a list of CSS selectors to identify the individual
-    columns within a row.
+    Parse HTML content into a tabular format.
+
+    This function supports two modes:
+    1. Headless browser mode: Creates a new Playwright browser instance to parse
+       HTML from files or content strings, optionally saving results to JSON
+    2. Live page mode: Uses an existing live browser page for parsing without
+       creating a new browser instance (more efficient for real-time data extraction)
+
+    Supports two extraction methods (configured in schema):
+    1. "locator" (default): Uses Playwright locators with auto-wait
+    2. "server_side": Parses HTML server-side
+
+    Args:
+        brand_id: Brand identifier for custom parsing functions
+        schema: Schema definition with CSS selectors and extraction_method configuration
+        bundle_dir: Directory containing HTML files to parse (headless mode)
+        html_content: HTML content string to parse (headless mode)
+        page: Live Playwright page object to parse from (live mode)
+
+    Returns:
+        BundleOutput containing parsed data and metadata
+
+    Note: When using headless mode, exactly one of bundle_dir or html_content must be provided.
+          When using live mode, only the page parameter is needed.
+          Extraction method is configured via schema.extraction_method field.
     """
+    extraction_method = getattr(schema, "extraction_method", "locator")
+
     logger.info(
-        f"Parsing HTML content: {html_content[:200] if html_content else 'None'}",
-        extra={"schema": schema},
+        f"Parsing HTML content: {html_content[:200] if html_content else 'None'} with page: {page is not None}",
+        extra={"schema": schema, "extraction_method": extraction_method},
     )
-    if not (bundle_dir is None) ^ (html_content is None):
-        raise ValueError("Exactly one of bundle_dir or html_content must be provided")
 
     data: list[dict[str, str | list[str]]] = []
-    async with async_playwright() as pw:
-        browser = await pw.chromium.launch()
-        context = await browser.new_context()
-        context.set_default_timeout(TIMEOUT)
-        context.set_default_navigation_timeout(TIMEOUT)
-        page = await context.new_page()
+
+    if extraction_method == "server_side":
```

Comment:
I found the name is a bit inconsistent. `locator` (n), `server_side` (n), and `evaluate` (v). Maybe we should come up with better name?